text/0000-crates-io-token-scopes.md

```diff
@@ -0,0 +1,230 @@
+# crates.io token scopes RFC
+
+- Feature Name: `crates_io_token_scopes`
+- Start Date: (fill me in with today's date, YYYY-MM-DD)
+- RFC PR: [rust-lang/rfcs#0000](https://github.com/rust-lang/rfcs/pull/0000)
+- crates.io issue: [rust-lang/crates.io#0000](https://github.com/rust-lang/crates.io/issues/0000)
+
+# Summary
+[summary]: #summary
+
+This RFC proposes implementing scopes for crates.io tokens, allowing users to
+choose which endpoints the token is allowed to call and which crates it's
+allowed to affect.
+
+# Motivation
+[motivation]: #motivation
+
+While the current implementation of API tokens for crates.io works fine for
+developers using the `cargo` CLI on their workstations, it's not acceptable for
+CI scenarios, such as publishing crates automatically once a git tag is pushed.
+
+The implementation of scoped tokens would allow users to restrict the actions a
+token can do, decreasing the risk in case of automation bugs or token
+compromise.
+
+# Guide-level explanation
+[guide-level-explanation]: #guide-level-explanation
+
+During token creation, the user will be prompted to select the permissions the
+token will have. Two sets of independent scopes will be available: the
+endpoints the token is authorized to call, and the crates the token is allowed
+to act on.
+
+## Endpoint scopes
+
+The user will be able to choose one or more endpoint scopes. This RFC proposes
+adding the following endpoint scopes:
+
+* **publish-new**: allows publishing new crates
+* **publish-update**: allows publishing a new version for existing crates the
+  user owns
+* **yank**: allows yanking and unyanking existing versions of the user's crates
+* **change-owners**: allows inviting new owners or removing existing owners
+* **legacy**: allows accessing all the endpoints on crates.io except for
+  creating new tokens, like tokens created befores the implementation of this
+  RFC.
+
+More endpoint scopes might be added in the future without the need of a
+dedicated RFC.
+
+The crates.io UI will pre-select the scopes needed by the `cargo` CLI, which at
+the time of writing this RFC are `publish-new`, `publish-update`, `yank` and
+`change-owners`. The user will have to explicitly opt into extra scopes or the
+legacy permission model.
+
+Tokens created before the implementation of this RFC will default to the legacy
+scope.
+
+## Crates scope
+
+The user will be able to opt into limiting which crates the token can act on by
+defining a crates scope.
+
+The crates scope can be left empty to allow the token to act on all the crates
```

Comment:
If we support `*` then shouldn't existing tokens default to the wildcard instead of an empty string?

Reply:
It's probably worth differentiating between a legacy token and a new `*` token, if only because then it'll be *possible* to revoke all legacy tokens while ignoring newly generated personal wildcard tokens.

Reply:
we can still do the same thing by looking at the timestamp of when the token was created :)